### ISSUE TITLE
Server idle timeout does not cancel a servers outgoing Source

### ISSUE THREAD

**Issue by jroper** [akka/akka-grpc#664]
I've only tried to reproduce this issue in akka-grpc, but the issue could well be in akka-http, I haven't attempted to remove akka-grpc from the equation (mostly because I don't have a generic HTTP/2 client on hand to test with). I've tested with akka-grpc 0.6.0 and akka-http 10.1.9.

With the following protobuf:

```protobuf
syntax = "proto3";
message StreamedIn {}
message StreamedOut {
    string message = 1;
}
service Streamed {
    rpc Call(StreamedIn) returns (stream StreamedOut);
}
```

And the following code:

```scala
import akka.NotUsed
import akka.actor.ActorSystem
import akka.grpc.GrpcClientSettings
import akka.http.scaladsl.{Http, HttpConnectionContext}
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.Source
import com.typesafe.config.ConfigFactory


object Test {

  def main(args: Array[String]) {
    val config = ConfigFactory
      .parseString(
        """
          |akka.http.server.preview.enable-http2 = on
          |akka.http.server.idle-timeout = 2s
        """.stripMargin)
      .withFallback(ConfigFactory.defaultReference())

    implicit val system = ActorSystem("Test", config)
    implicit val materializer = ActorMaterializer()
    // needed for the future map/flatmap in the end
    implicit val executionContext = system.dispatcher

    val streamedServer = new Streamed {
      override def call(in: StreamedIn): Source[StreamedOut, NotUsed] =
        Source.single(StreamedOut("After this the stream will pause indefinitely"))
          .concat(Source.maybe)
          .watchTermination() { (_, terminated) =>
            terminated.onComplete { result =>
              println(s"Server received stream terminated: $result")
            }
            NotUsed
          }
    }


    // Bind service handler servers to localhost:8080/8081
    val binding = Http().bindAndHandleAsync(
      StreamedHandler(streamedServer),
      interface = "127.0.0.1",
      port = 8080,
      connectionContext = HttpConnectionContext())

    // report successful binding
    binding.foreach { binding =>
      println(s"gRPC server bound to: ${binding.localAddress}")

      val client = StreamedClient(GrpcClientSettings.connectToServiceAt("127.0.0.1", 8080).withTls(false))
      client.call(StreamedIn()).runForeach(msg => println(s"Client received message: $msg")).onComplete { result =>
        println(s"Client received stream terminated: $result")
      }
    }
  }
}
```

I see the following output:

```
gRPC server bound to: /127.0.0.1:8080
Client received message: StreamedOut(After this the stream will pause indefinitely)
Client received stream terminated: Failure(io.grpc.StatusRuntimeException: UNAVAILABLE: io exception)
```

The server correctly closes the connection after the configured 2 second idle timeout, the client stream sees that, but the server stream doesn't.

**Comment by jroper** [comment]
This seems to also happen when the client closes the connection without gracefully terminating the HTTP/2 streams, or if the connection fails for some other reason.

**Comment by viktorklang** [comment]
Adding more debug information:

```
val client = StreamedClient(GrpcClientSettings.connectToServiceAt("127.0.0.1", 8080).withTls(false))
      client.call(StreamedIn()).runForeach(msg => println(s"Client received message: $msg")).onComplete { result =>
        println(s"Client received stream terminated: $result")
        client.close() <-<-<-< ADDING THIS DOES *NOT* CAUSE THE `watchTermination` TO TRIGGER
        materializer.shutdown() <-<-<-< ADDING THIS *DOES* CAUSE THE `watchTermination` TO TRIGGER
      }
```

IMO: closing the client should cause the watchTermination to trigger, but it currently *doesn't*.

**Comment by viktorklang** [comment]
Adding a deadline to the gRPC client configuration closes the connection after the given time:

```
      val client =
        StreamedClient(GrpcClientSettings.connectToServiceAt("127.0.0.1", 8080).withTls(false).withDeadline(2.seconds))
      client.call(StreamedIn()).runForeach(msg => println(s"Client received message: $msg")).onComplete { result =>
        println(s"Client received stream terminated: $result")
      }
```

Output:
```
scala> gRPC server bound to: /127.0.0.1:8080
Client received message: StreamedOut(After this the stream will pause indefinitely)
Client received stream terminated: Failure(io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: deadline exceeded after 1965801314ns)
Server received stream terminated: Success(Done)
```

**Comment by viktorklang** [comment]
Diagnosis:

I believe the patient's problems stem from a discrepancy between handling of idleTimeout between HTTP and HTTP2 support in akka-http.

In Http the Tcp server sets an Inf idleTimeout on itself:

```scala
    Tcp()
      .bind(
        interface,
        port,
        settings.backlog,
        settings.socketOptions,
        halfClose = false,
        idleTimeout = Duration.Inf // we knowingly disable idle-timeout on TCP level, as we handle it explicitly in Akka HTTP itself
      )
```

Then it uses `settings.idleTimeout` at the Http layer like so:

```scala
  private def fuseServerBidiFlow(
    settings:          ServerSettings,
    connectionContext: ConnectionContext,
    log:               LoggingAdapter): ServerLayerBidiFlow = {
    val httpLayer = serverLayerImpl(settings, None, log, connectionContext.isSecure)
    val tlsStage = sslTlsStage(connectionContext, Server)

    val serverBidiFlow =
      settings.idleTimeout match {
        case t: FiniteDuration => httpLayer atop tlsStage atop HttpConnectionIdleTimeoutBidi(t, None)
        case _                 => httpLayer atop tlsStage
      }

    GracefulTerminatorStage(system, settings) atop serverBidiFlow
  }
```

For Http2 it is very different, it sets the `settings.idleTimeout` on the TCP server itself:

```scala
Tcp().bind(interface, effectivePort, settings.backlog, settings.socketOptions, halfClose = false, settings.timeouts.idleTimeout)
```

And then does not further use the `settings.idleTimeout`.


Therefor I posit that what needs to be done is to use `settings.idleTimeout` in the same fashion for Http2 as for Http—by adding a HttpConnectionIdleTimeoutBidi to the mix per SubStream?

Now, where to best add this?

**Comment by viktorklang** [comment]
@jroper @jrudolph I haven't had a chance to test this out yet, but since it is getting to EoD Friday for me, have a look at this potential fix when/if you have time: https://github.com/akka/akka-http/tree/wip-http2-idle-timeout-%E2%88%9A

**Comment by jrudolph** [comment]
IIRC the only reason we don't use the tcp-level one is to be able to change the error message. So, it's somewhat weird that it wouldn't work functionally even with just the TCP timeout. So, while the fix might be helping for that particular case, there might still be another problem that the HTTP/2 implementation doesn't react properly to TCP-level problems.

**Comment by viktorklang** [comment]
@jrudolph True, OTOH I think the fix is rather solid since it uses the same mechanism as for HTTP1. Perhaps open a separate issue to investigate TCP handling for HTTP2?

**Comment by jrudolph** [comment]
Thanks a lot for the reproducers and discussion. I found the root cause here: https://github.com/akka/akka-http/issues/2777.

**Comment by raboof** [comment]
Since the upstream issue was fixed in Akka HTTP 10.1.11 and Akka gRPC is on Akka HTTP is already on 10.2.x I think we can close this
